kedro/ipython/__init__.py

```diff
@@ -21,46 +20,71 @@
 from kedro.framework.startup import _is_project, bootstrap_project
 
 logger = logging.getLogger(__name__)
-default_project_path = Path.cwd()
 
 
-def _remove_cached_modules(package_name):  # pragma: no cover
-    to_remove = [mod for mod in sys.modules if mod.startswith(package_name)]
-    # `del` is used instead of `reload()` because: If the new version of a module does not
-    # define a name that was defined by the old version, the old definition remains.
-    for module in to_remove:
-        del sys.modules[module]
+def load_ipython_extension(ipython):
+    """
+    Main entry point when %load_ext kedro.ipython is executed, either manually or
+    automatically through `kedro ipython` or `kedro jupyter lab/notebook`.
+    IPython will look for this function specifically.
+    See https://ipython.readthedocs.io/en/stable/config/extensions/index.html
+    """
+    if _find_kedro_project(Path.cwd()) is None:
+        logger.warning(
+            "Kedro extension was registered but couldn't find a Kedro project. "
+            "Make sure you run '%reload_kedro <project_root>'."
+        )
+        return
 
+    ipython.register_magic_function(magic_reload_kedro, magic_name="reload_kedro")
+    reload_kedro()
 
-def _find_kedro_project(current_dir: Path):  # pragma: no cover
-    while current_dir != current_dir.parent:
-        if _is_project(current_dir):
-            return current_dir
-        current_dir = current_dir.parent
 
-    return None
+@needs_local_scope
+@magic_arguments()
+@argument(
+    "path",
+    type=str,
+    help=(
+        "Path to the project root directory. If not given, use the previously set"
+        "project root."
+    ),
+    nargs="?",
+    default=None,
+)
+@argument("-e", "--env", type=str, default=None, help=ENV_HELP)
+@argument(
+    "--params",
+    type=lambda value: _split_params(None, None, value),
+    default=None,
+    help=PARAMS_ARG_HELP,
+)
+def magic_reload_kedro(line: str, local_ns: Dict[str, Any] = None):
+    """
+    The `%reload_kedro` IPython line magic. See
+     https://kedro.readthedocs.io/en/stable/tools_integration/ipython.html for more.
+    """
+    args = parse_argstring(magic_reload_kedro, line)
+    reload_kedro(args.path, args.env, args.params, local_ns)
 
 
 def reload_kedro(
-    path: str = None, env: str = None, extra_params: Dict[str, Any] = None
-):  # pragma: no cover
+    path: str = None,
+    env: str = None,
+    extra_params: Dict[str, Any] = None,
+    local_ns: Optional[Dict[str, Any]] = None,
```

Comment:
What is `local_ns`? I'd be in favour of having longer argument names if that means it's more descriptive.

Reply:
`local_ns` is the argument that IPython passes to magics that are decorated with `needs_local_scope`, which is the feature that allows this implementation to work. `local_ns` is a dict containing all declarations in the scope of the IPython session and is persisted between calls to magics / extensions being reloaded. Unfortunately, the name must be `local_ns` in the function signature of the magic for IPython to use it correctly. This feature of IPython seems poorly documented, or I would share some docs.

That being said, we should use a more descriptive name elsewhere -- `local_namespace`. I will also change the `project_path` key of `local_ns` to  `_project_path`, to avoid users accidentally overwriting this.

These changes are in [838aaf1](https://github.com/kedro-org/kedro/pull/1955/commits/838aaf1bb7e8251e45e0794991151fc34ac5fb91)

Reply:
Thanks for explaining @jmholzer, this makes sense now.